### PR TITLE
Date formatting

### PR DESCRIPTION
**Pull request by tomberek**
Fixes the issue talked about at: http://stackoverflow.com/questions/7556591/javascript-date-object-always-one-day-off?noredirect=1&lq=1

When picking a date (as opposed to a date/time) almost no one will want 29 Jan to be converted to 28 Jan GMT-5.

### CODE REVIEW

**Comment by tomberek** [comment]
This is probably too heavy handed, but something along these lines should be available to allow users to choose a date without having to be concerned by locales/time zones/etc.

**Comment by jpetitcolas** [comment]
In my opinion this case should not be handled directly in admin-on-rest. Indeed, it is recommended to always return a timezone with your dates in API to prevent these kinds of issues. If you don't, I would instead recommend to add an interceptor in your REST client to make these conversions.

Moreover, what about if you want to display date in local browser time? :)

**Comment by tomberek** [comment]
All true. But in this case admin-on-rest is taking the raw data returned by REST and sending it through `new Date(value)`, then displaying the result, potentially distorting the value. In the case where I came across this, my DB was returning a value `2017-01-30` which `Date` interpreted as `2017-01-29 19:00 GMT -5` which then got displayed by admin-rest as `2017-01-29`, truncating the time. So each time you re-rendered the component, the date dropped by one day, obviously not correct.

Due to the quirk of how Date parses dates, it won't occur with `2017/01/30`, so that's where my hacky solution came from. Would it be better for admin-on-rest to not even run the date through `Date`? Or is there some other way to enforce that a Date is "timezone-less" and thus should not be corrected for or interpreted as time-zone dependent? An example is a birthday. Most people interpret it as just a constant day, vice correcting for timezone with the true time of birth or international dateline, etc.

**Comment by fzaninotto** [comment]
Your solution removes hours and seconds from Datetime, it it's not good. The `DateField` has a `withTime` param that you should take into account.

I think the problem isn't with the parsing of the date, it's with the rendering of the dates.

For the `DateField`, did you try passing `{ timeZone: 'America/New_York' }` as `options` prop? This forces the timezone for the locale date conversion.

https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/Date/toLocaleString

**Comment by tomberek** [comment]
@fzaninotto : Yes, I've tried timeZone and locale. I suspect a modified DateTimeFormat could probably do the trick.

I agree that my solution in this PR is __not good__ for the general case. But I do think the datify function needs a modification. new Date is manipulating the input as a moment in time, then truncating the time for display. 

Here's the note from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date regarding the use of `new Date`

```
Note: parsing of date strings with the Date constructor (and Date.parse, they are equivalent) is
 strongly discouraged due to browser differences and inconsistencies. Support for RFC 2822
 format strings is by convention only. Support for ISO 8601 formats differs in that 
date-only strings (e.g. "1970-01-01") are treated as UTC, not local.
```

Some other libraries solve the problem by internally replacing `-` by `/`. This also works for your library, (it was a previous attempt at a PR, not sure if you'd find that acceptable).

https://github.com/danvk/dygraphs/blob/3f675fe524eece31f39fb68601d2d2d7d5720941/dygraph-utils.js#L510

**Comment by jpetitcolas** [comment]
Replacing `-` by `/` looks like a better idea in this case. Yet, we would need to test this method on several timezones and browsers (especially Safari, which has some troubles with dates if I remember correctly. @Kmaschta or @ThieryMichel, can you confirm?).

File: src/mui/input/DateInput.js [hunk 1]
```diff
@@ -7,8 +7,7 @@ export const datify = input => {
         return null;
     }
 
-    let date = input instanceof Date ? input : new Date(input);
-	date = new Date(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate());
+    let date = input instanceof Date ? input : new Date(input.replace('-','/','g'));
```
Comment (fzaninotto) [inline src/mui/input/DateInput.js]: nice, but you should do that in the other place, too. And switch back to `const`

**Comment by leesei** [comment]
@tomberek Any update on this PR?

**Comment by tomberek** [comment]
@leesei - No.
I am still personally interested, but my project has moved on to a different solution.

**Comment by leesei** [comment]
I see. I'll be using `<DateInput>` in my project.
Would you mind me creating a PR based on code in this PR if I see any problem?

**Comment by tomberek** [comment]
go ahead

**Comment by leesei** [comment]
On second thought `<DateInput>` is not wrong in returning `Date` object of UTC time, which is the internal representation of all times in JavaScript VM.
If the issue is transforming the date string "YYYY-MM-DD" (presumed local time) to `Date` object, do refer to `dateFormatter` in #555:

```js
const dateFormatter = v => {
  // v is a string of "YYYY-MM-DD" format
  const match = /(\d{4})-(\d{2})-(\d{2})/.exec(v);
  if (match === null) return;
  const d = new Date(match[1], parseInt(match[2])-1, match[3]);
  if (isNaN(d)) return;
  return d;
};
```

**Comment by fzaninotto** [comment]
Agreed, closing.

**Comment by WesleyKapow** [comment]
Just hit this, and had to use the following to get `DateField` rendering correctly:
```js
<DateField source="birthDate" options={{ timeZone: "UTC" }} />
```

**Pull request closed without merging**